### PR TITLE
[ROCm] Add hipDNN backend support for convolution

### PR DESCRIPTION
**Pull request by zjgarvey**
This PR adds [hipDNN](https://github.com/ROCm/rocm-libraries/tree/develop/projects/hipdnn) convolution support to PyTorch on ROCm, following the newer dispatch mechanism which no longer requires per-backend frontend op registration.

New ROCm kernel providers will be able to register plugins and ship faster kernels through hipDNN without touching PyTorch code. The graph-based API of hipDNN also supports fused kernels. The eventual goal is to replace all MIOpen-specific PyTorch code with hipDNN, wherein MIOpen will be just one of multiple first-class kernel providers.

# Note

This PR is being opened parallel to #177534 in anticipation that batch norm's dispatch migration issues might stall the review process for hipdnn as a whole. 

# Summary

- Integrates the lightweight hipDNN frontend graph API as an alternative convolution backend to MIOpen on ROCm, covering forward, backward-data, and backward-weight for 2D/3D convolutions including transposed variants.

- Adds cmake `USE_HIPDNN` detection, `torch.backends.hipdnn` Python module with `enabled`/`flags()` context manager, and `compiledWithHipDNN()` runtime hook.     

- Implements a generic thread-local `ParamsLRUCache` to amortize hipDNN `graph->build()` cost across repeated convolution calls with identical parameters. 

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01

### CODE REVIEW

**Review by jeffdaily — CHANGES_REQUESTED** [review]
My only issue is having aten/src/ATen/native/utils/ParamsLRUCache.h as a generic ParamsLRUCache for pytorch.  It's not a bad idea, and this implementation seems inspired by the BenchmarkCache in aten/src/ATen/native/cudnn/Conv_v8.cpp.  But I would defer such an improvement to a future PR where any/all LRU cache implementations within pytorch utilize a common implementation.  So please move the ParamsLRUCache implementation into aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp since that's the only place it is used currently.  This will also prevent exposing this struct as a new public API header.

**Review by jeffdaily — DISMISSED** [review]

File: aten/src/ATen/cuda/detail/CUDAHooks.h [hunk 1]
```diff
@@ -42,6 +42,7 @@ struct CUDAHooks : public at::CUDAHooksInterface {
   Allocator* getPinnedMemoryAllocator() const override;
   bool compiledWithCuDNN() const override;
   bool compiledWithMIOpen() const override;
+  bool compiledWithHipDNN() const override;
```
Comment (malfet) [inline aten/src/ATen/cuda/detail/CUDAHooks.h:45]: Just curios, why do we need to go thru CUDAHooks, why not just call it directly?

File: aten/src/ATen/hipdnn/Handle.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include <c10/hip/HIPStream.h>
+
+#include <ATen/hip/Exceptions.h>
+#include <ATen/hipdnn/Exceptions.h>
+#include <hipdnn/frontend/hipdnn_frontend/Handle.hpp>
+
+namespace at::native {
+namespace {
+
+void createHipdnnHandle(hipdnnHandle_t* handle) {
```
Comment (malfet) [inline aten/src/ATen/hipdnn/Handle.cpp:12]: This looks like a sed -e `s/Cudnn/Hipdnn/ aten/src/ATen/cudnn/Handle.cpp`

See https://github.com/pytorch/pytorch/blob/f00374eb7413c40e3f61a1f6f889fd210e28bb23/aten/src/ATen/cudnn/Handle.cpp#L10

Why can't it be hipified then?

File: test/nn/test_convolution.py [hunk 3]
```diff
@@ -4348,6 +4350,210 @@ def test_depthwise_conv_64bit_indexing(self, device):
...
+                ((2, 128, 32, 32), (128, 1, 3, 3), 1, 1, 1, 128),
+                name="depthwise",
+            ),
+            subtest(
+                ((1, 3, 224, 224), (64, 3, 7, 7), 2, 3, 1, 1),
+                name="resnet_first",
+            ),
+        ],
+    )
+    def test_conv2d_hipdnn_configs(self, device, config):
```
Comment (malfet) [inline test/nn/test_convolution.py]: Why this needs to be a separate test rather than reusing existing one?

Or why one needs to run it specifically for hipDNN, but not for cuDNN? Or for any other convolution engine?

File: torch/backends/hipdnn/__init__.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# mypy: allow-untyped-defs
+import sys
```
Comment (malfet) [inline torch/backends/hipdnn/__init__.py:2]: Again, this looks almost like a verbatim copy of `torch/backends/cudnn/__init__.py`

**Review by malfet — CHANGES_REQUESTED** [review]
It looks like this PR does a lot of copy-n-paste from cudnn into hipdnn on the API level, but doesn't explain much what are the differentiation between two approaches.

I know there could be some sensibility about naming, but as PyTorch is not compilable with both CuDNN and HipDNN backends at the same time, should later just reuse scaffolding from the former rather than write a replica of one?

cc: @albanD who is driving more `torch.accelerator` generalization

**Comment by malfet** [comment]
> Is there a public buck-related file we can update to set AT_HIPDNN_ENABLED to false until Meta is ready to enable it internally?

None at the moment

File: aten/src/ATen/hipdnn/Handle.cpp (hunk 2 above)
Comment (zjgarvey) [reply aten/src/ATen/hipdnn/Handle.cpp:12]: I'll address the primary comment here as well, since it's all somewhat related. 

hipDNN is not the same thing as "hipified" cuDNN.

We aren't guaranteeing that there is a 1-1 mapping from cudnn frontend to hipdnn frontend. They are similar sounding because they both provide a graph api for specifying and launching fused kernels.

For AMD/ROCm, hipDNN is intended to really to expand and replace our pytorch integration surface area beyond MIOpen, not to copy/paste from cuDNN. If there is a new path for adding convolution backends through, e.g., `torch.accelerator`, I'd be happy to help move in that direction, too.

File: test/nn/test_convolution.py (hunk 3 above)
Comment (zjgarvey) [reply test/nn/test_convolution.py]: hipdnn is currently opt-out by default, and needs to be enabled by a flag. These tests are hipdnn only because they call `torch.backends.hipdnn.flags(enabled=True)`. If there is a good way to add a new parameter to existing conv tests to enable hipdnn when available, I'd be okay with that.

File: aten/src/ATen/cuda/detail/CUDAHooks.h (hunk 1 above)
Comment (zjgarvey) [reply aten/src/ATen/cuda/detail/CUDAHooks.h:45]: My understanding is that CUDAHooks are for exposing GPU-specific build config info to non-GPU specific files like `pytorch/aten/src/ATen/native/Convolution.cpp` where this is primarily used. We don't pass `-DUSE_HIPDNN` when building that file, for example.

File: aten/src/ATen/cuda/detail/CUDAHooks.cpp [hunk 5]
```diff
@@ -280,6 +280,10 @@ bool CUDAHooks::compiledWithMIOpen() const {
   return AT_ROCM_ENABLED();
 }
 
+bool CUDAHooks::compiledWithHipDNN() const {
+  return AT_HIPDNN_ENABLED();
```
Comment (zjgarvey) [inline aten/src/ATen/cuda/detail/CUDAHooks.cpp]: @malfet @jeffdaily would it be acceptable to remove the macro and change this to something like the following?

```cpp
bool CUDAHooks::compiledWithHipDNN() const {                                                                                                      
  #ifdef USE_HIPDNN                                                                                                                                 
    return true;
  #else                                                                                                                                             
    return false; 
  #endif                                                                                                                                            
  }
```

We don't use the macro anywhere else but here.

**Review by jeffdaily — APPROVED** [review]

**Comment by zjgarvey** [comment]
@malfet Can you let me know if the suggested fix is acceptable to you? I'm not sure how to push this along further and would appreciate some help.

File: aten/src/ATen/hipdnn/Handle.cpp (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/hipdnn/Handle.cpp:12]: So, will it be used in conjunction with MiOpen or will it replace MiOpen? I don't think it's healthy to have one more backend which is mutually exclusive. I don't mind renaming miopen folder into rocdnn and keep both old and new implementation here, but I really don't see that creating another copy of the same boilerplate code helps

Comment (zjgarvey) [reply aten/src/ATen/hipdnn/Handle.cpp:12]: > So, will it be used in conjunction with MIOpen or will it replace MIOpen?

hipDNN will eventually replace MIOpen. In the medium term they need to coexist while (a) we bring hipDNN backend coverage online op-by-op (conv here, SDPA/batch norm to follow) and (b) hipDNN itself matures and more kernel providers ship through it. Once hipDNN reaches parity, `aten/src/ATen/miopen/` gets removed outright.

> I don't think it's healthy to have one more backend which is mutually exclusive. I don't mind renaming miopen folder into rocdnn and keep both old and new implementation here, but I really don't see that creating another copy of the same boilerplate code helps

Yeah, `Handle.cpp` is something of a find/replace of miopen or cudnn, both of which have similar handle lifecycles.

- `Exceptions.h` is quite different. hipDNN uses `TORCH_CHECK_WITH`/`HipDNNError` with a separate `HIPDNN_FE_CHECK` for the frontend's status type; MIOpen still uses the older `std::runtime_error` subclass.
- `Utils.h` has a `TensorAttributes` builder for hipDNN vs. MIOpen's zero-stride contiguity helper.
- The main reason I kept these separate is because the include surface differs: MIOpen is a C API; hipDNN frontend is a C++ graph API with its own headers (`hipdnn_frontend.hpp` etc.) that aren't guaranteed to be present at configure time.

So the real duplication is only~50 lines of handle-pool boilerplate plus a one-line umbrella header.

------

The options I see:

1. **Share boilerplate, rename the folder to something "rocm" generic.** Shared files need conditional includes since the hipDNN headers may or may not be present, which makes them uglier. The backend rename itself is cumbersome, and when we eventually drop MIOpen we can't just delete a standalone folder.

2. **Keep folders separate (current PR).** Small amount of duplicated boilerplate in `Handle.cpp` and the wrapper header; everything else genuinely differs. The `hipdnn/` folder is not compiled at all unless hipDNN is found at configure time, and MIOpen removal later is a clean folder deletion.

3. **Hipify from the cuDNN sources (your original suggestion).**  hipDNN is not a 1-1 mirror of cuDNN. The graph APIs are superficially similar because both are graph-based, but the semantics and extension points differ. A mechanical Python rewrite of cuDNN sources would tie hipDNN's evolution to cuDNN's design decisions, which isn't a good foundation I want to build on for a library we're actively developing.

My preference is (2) for the reasons above, but I'm open to (1) if you feel the boilerplate cost strongly outweighs the eventual-cleanup cost.

File: torch/testing/_internal/common_utils.py [hunk 6]
```diff
@@ -1623,6 +1623,24 @@ def get_all_dsls() -> list[str]:
 # TODO: Remove PYTORCH_MIOPEN_SUGGEST_NHWC once ROCm officially supports NHWC in MIOpen
 # See #64427
 TEST_WITH_MIOPEN_SUGGEST_NHWC = os.getenv('PYTORCH_MIOPEN_SUGGEST_NHWC', '0') == '1'
+
+# TODO: Remove PYTORCH_TEST_WITH_HIPDNN once hipdnn is enabled by default on ROCm.
+# Opt-in: enable hipDNN globally for the test session. Existing CUDA conv tests
+# (numerics, gradcheck, channels_last, decomp, OpInfo) then run against hipdnn
+# instead of MIOpen on ROCm without per-test changes. Activation happens in
+# TestCase.run() via torch.backends.hipdnn.flags(), since direct mutation of the
+# global flag is forbidden after disable_global_flags() above.
```
Comment (zjgarvey) [inline torch/testing/_internal/common_utils.py]: remove this part of the comment.

File: torch/csrc/Module.cpp [hunk 7]
```diff
@@ -1358,6 +1358,29 @@ static PyObject* THPModule_immediateMiopen(
   Py_RETURN_FALSE;
 }
 
+static PyObject* THPModule_setUserEnabledHipdnn(
```
Comment (zjgarvey) [inline torch/csrc/Module.cpp]: look into whether it is better to do this via pybind.

**Comment by yanjianglu** [comment]
May I ask if this feature can still be merged?

**Comment by zjgarvey** [comment]
> May I ask if this feature can still be merged?

Hey, thanks for asking. Unfortunately, we have decided to close this pull request. The hipDNN team has decided to go a different direction with their integration efforts. 

If anyone wants to revisit the approach presented here, we will keep an archive branch for this at https://github.com/ROCm/pytorch/tree/hipdnn_develop (has hipDNN support for batch norm, conv, attention).
